content/blog/2024/09/camel-ai-examples/index.md

```diff
@@ -0,0 +1,296 @@
+import ai.djl.modality.Classifications;
+import ai.djl.modality.cv.Image;
+import net.sf.extjwnl.data.POS;
+import net.sf.extjwnl.data.PointerUtils;
+import net.sf.extjwnl.dictionary.Dictionary;
+
+public class image_to_text extends RouteBuilder {
+
+    @Override
+    public void configure() throws Exception {
+        from("file:data/inbox?include=.*\\.(jpg|png)")    // (1)
+            .log("Processing: ${headers.camelFileName}")
+            .to("djl:cv/object_detection?artifactId=ssd") // (2)
+            .convertBodyTo(Image[].class)                 // (3)
+            .split(body())                                           // (4)
+                .to("djl:cv/image_classification?artifactId=resnet") // (4)
+                /*
+                 * The output from the image classification model is classified
+                 * as one of 1000 labels from WordNet.
+                 * Since it's too fine-grained, we want to find the higher-level
+                 * group (= hypernym) for the classification using the WordNet
+                 * dictionary.
+                 */
+                .process(this::extractClassName)          // (5)
+                .process(this::addHypernym)               // (5)
+                .log("  => ${body}");                     // (6)
+    }
+
+    void extractClassName(Exchange exchange) {
+        var body = exchange.getMessage().getBody(Classifications.class);
+        var className = body.best().getClassName().split(",")[0].split(" ", 2)[1];
+        exchange.getMessage().setBody(className);
+    }
+
+    void addHypernym(Exchange exchange) throws Exception {
+        var className = exchange.getMessage().getBody(String.class);
+        var dic = Dictionary.getDefaultResourceInstance();
+        var word = dic.getIndexWord(POS.NOUN, className);
+        if (word == null) {
+            throw new RuntimeCamelException("Word not found: " + className);
+        }
+        var hypernyms = PointerUtils.getDirectHypernyms(word.getSenses().get(0));
+        var hypernym = hypernyms.stream()
+                .map(h -> h.getSynset().getWords().get(0).getLemma())
+                .findFirst().orElse(className);
+        exchange.getMessage().setBody(List.of(className, hypernym));
+    }
+}
+```
+
+What the code does is as follows:
+
+1. Read image files (`.jpg` or `.png`) from the `data/inbox` directory
+2. Feed them to a pre-trained SSD model for object detection
+3. Convert the results of type `DetectedObjects` to `Image[]`
+4. Split the `Image[]` data one by one and feed each into a pre-trained ResNet model for image classification
+5. Look up the hypernym of each classification result with the WordNet database
+6. Output the final results to log (or send them as a message to the central server in a real scenario)
+
+### Image to Text: Running the example
+
+The code can be run with [Camel JBang](/manual/camel-jbang.html):
+
+```console
+camel run --camel-version=4.8.0-SNAPSHOT image_to_text.java
```

Comment:
Right. I'd keep it as is as it's intended to be published before 4.8.0 release for historical reasons.

Reply:
OK, I added how to run post 4.8.0 following your feedback. That should be more reader-friendly. Thanks Amos.

Reply:
OK, it is a very impress example. Is there github repo to host these examples? I'm looking forward to trying but it looks like no such images files, like `doge.jp`?

Reply:
Thanks, this is intended to be the place for the examples: https://github.com/megacamelus/camel-ai-examples
I excluded `doge.jpg` because I don't think it can be published under Apache-2.0. You can test the example with any funny images of your choice ;-)

Reply:
Nice work!

Reply:
BTW, that megacamelus repository has a few images: https://github.com/megacamelus/camel-ai-examples/tree/main/data/objects which are picked up from the DJL project itself (Apache-2.0).
You can use them to test the example.